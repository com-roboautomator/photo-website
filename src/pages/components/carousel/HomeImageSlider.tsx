import {createStyles, withStyles} from '@material-ui/core'
import {
    AbstractImageSliderProps,
    AbstractImageSliderState,
} from './AbstractImageSlider'
import AbstractImageSlider from './AbstractImageSlider'
import {collection} from '../../../assets/data/ImageDataStructure'
import HomeImageSliderButtonWrapper from './HomeImageSliderButtonWrapper'
import HomeCollectionCard from '../card/HomeCollectionCard'

interface HomeImageSliderProps extends AbstractImageSliderProps {}
interface HomeImageSliderState extends AbstractImageSliderState {}

class ImageSlider extends AbstractImageSlider<
    HomeImageSliderProps,
    HomeImageSliderState
> {
    constructor(props: HomeImageSliderProps) {
        super(props)
        this.state = {
            selected: this.props.startingIndex ? this.props.startingIndex : 0,
            properties: this.props.data,
            property: this.props.data[
                this.props.startingIndex ? this.props.startingIndex : 0
            ],
        }
    }

    render() {
        const classes = this.props.classes
        const {properties, property} = this.state

        for (let x = 0; x < properties.length; x++) {
            properties[x].index = x
        }

        return (
            <main className={classes.container}>
                <div
                    data-testid="ImageSlider-Container"
                    className={classes.wrapper}>
                    <HomeImageSliderButtonWrapper
                        index={property.index}
                        length={properties.length}
                        next={this.next}
                        previous={this.previous}
                    />
                    <div
                        data-testid="ImageSlider-Card-Slider"
                        className={classes.card_slider}
                        style={{height: `${this.props.height}px`}}
                        onWheel={(e) => {
                            this.handleWheelMovement(e)
                        }}>
                        <div
                            className={classes.card_slider_wrapper}
                            style={{
                                transform: `translateX(-${
                                    (property.index * 100) / properties.length
                                }%)`,
                            }}>
                            {properties.map((property: collection) => (
                                <HomeCollectionCard
                                    key={property.id}
                                    collection={property}
                                    height={
                                        this.props.height
                                            ? this.props.height
                                            : 200
                                    }
                                    selected={
                                        this.state.selected === property.index
                                    }
                                />
                            ))}
                        </div>
                    </div>
                </div>
            </main>
        )
    }
}

const styles = () =>
    createStyles({
        container: {
            display: 'flex',
            flexDirection: 'column',
            userSelect: 'none',
        },
        wrapper: {
            pointerEvents: 'auto',
            position: 'relative',
            display: 'flex',
            justifyContent: 'center',
            overflowX: 'hidden',
        },
        card_slider: {
            pointerEvents: 'auto',
            position: 'relative',
            maxWidth: '370px',
            width: '370px',
        },
        card_slider_wrapper: {
            pointerEvents: 'auto',
            display: 'flex',
            position: 'absolute',
            margin: '10px',
            transition: '300ms cubic-bezier(0.455, 0.03, 0.515, 0.955)',
        },
    })

export default withStyles(styles)(ImageSlider)
