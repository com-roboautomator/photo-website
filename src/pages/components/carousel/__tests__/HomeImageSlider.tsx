import {fireEvent, render} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'
import collection from 'src/assets/data/ImageDataStructure'
import HomeImageSlider from '../HomeImageSlider'

const coll: collection[] = [
    {
        key: 'first-test-collection-key',
        title: 'first-test-collection-title',
        index: 0,

        images: [
            {
                key: 'test-image-key',
                url: 'test-image-url',
                title: 'test-image-title',
                index: 0,
            },
        ],
    },
    {
        key: 'second-test-collection-key',
        title: 'second-test-collection-title',
        index: 1,

        images: [
            {
                key: 'test-image-key',
                url: 'test-image-url',
                title: 'test-image-title',
                index: 0,
            },
        ],
    },
    {
        key: 'third-test-collection-key',
        title: 'third-test-collection-title',
        index: 2,

        images: [
            {
                key: 'test-image-key',
                url: 'test-image-url',
                title: 'test-image-title',
                index: 0,
            },
        ],
    },
]

describe('Home Image Slider', () => {
    it('should match snapshot', () => {
        const ImageSliderRender = renderer
            .create(<HomeImageSlider data={coll} />)
            .toJSON()
        expect(ImageSliderRender).toMatchSnapshot()
    })

    it('should render correctly', () => {
        const {getByTestId} = render(<HomeImageSlider data={coll} />)

        expect(getByTestId('ImageSlider-Container')).toBeInTheDocument()
        expect(getByTestId('ImageSlider-Card-Slider')).toBeInTheDocument()
        expect(
            getByTestId('Home-Collection-Card-first-test-collection-title')
        ).toBeInTheDocument()
    })

    it('should increment image on button press', () => {
        const {getByTestId} = render(
            <HomeImageSlider data={coll} startingIndex={0} />
        )

        const firstImage = getByTestId(
            'Home-Collection-Card-first-test-collection-title'
        )
        const secondImage = getByTestId(
            'Home-Collection-Card-second-test-collection-title'
        )

        expect(firstImage).toHaveStyle({opacity: '100%', transform: 'scale(1)'})
        expect(secondImage).toHaveStyle({
            opacity: '90%',
            transform: 'scale(0.8)',
        })

        fireEvent.click(getByTestId('Arrow-Button-Right'))

        expect(firstImage).toHaveStyle({
            opacity: '90%',
            transform: 'scale(0.8)',
        })
        expect(secondImage).toHaveStyle({
            opacity: '100%',
            transform: 'scale(1)',
        })
    })

    it('should decrement image on button press', () => {
        const {getByTestId} = render(
            <HomeImageSlider data={coll} startingIndex={1} />
        )

        const firstImage = getByTestId(
            'Home-Collection-Card-first-test-collection-title'
        )
        const secondImage = getByTestId(
            'Home-Collection-Card-second-test-collection-title'
        )

        expect(firstImage).toHaveStyle({
            opacity: '90%',
            transform: 'scale(0.8)',
        })
        expect(secondImage).toHaveStyle({
            opacity: '100%',
            transform: 'scale(1)',
        })

        fireEvent.click(getByTestId('Arrow-Button-Left'))

        expect(firstImage).toHaveStyle({
            opacity: '100%',
            transform: 'scale(1)',
        })
        expect(secondImage).toHaveStyle({
            opacity: '90%',
            transform: 'scale(0.8)',
        })
    })

    it('should move on mouse scroll', () => {
        const {getByTestId} = render(
            <HomeImageSlider data={coll} startingIndex={0} />
        )

        const firstImage = getByTestId(
            'Home-Collection-Card-first-test-collection-title'
        )
        const secondImage = getByTestId(
            'Home-Collection-Card-second-test-collection-title'
        )

        expect(firstImage).toHaveStyle({opacity: '100%', transform: 'scale(1)'})
        expect(secondImage).toHaveStyle({
            opacity: '90%',
            transform: 'scale(0.8)',
        })

        fireEvent.wheel(getByTestId('ImageSlider-Card-Slider'), {
            deltaY: 30,
        })

        expect(firstImage).toHaveStyle({
            opacity: '90%',
            transform: 'scale(0.8)',
        })
        expect(secondImage).toHaveStyle({
            opacity: '100%',
            transform: 'scale(1)',
        })
    })
})
