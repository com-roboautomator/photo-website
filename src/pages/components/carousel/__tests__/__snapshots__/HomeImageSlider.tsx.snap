// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Image Slider should match snapshot 1`] = `
<main
  className="ImageSlider-container-1"
>
  <div
    className="ImageSlider-wrapper-2"
    data-testid="ImageSlider-Container"
  >
    <div
      className="HomeImageSliderButtonWrapper-button_wrapper-5"
      data-testid="ImageSlider-button-wrapper"
      style={
        Object {
          "background": "linear-gradient(to right, white, transparent 30%), linear-gradient(to left, white, transparent 30%)",
        }
      }
    >
      <div
        className="HomeImageSliderButtonWrapper-button_wrapper-5"
        data-testid="Buttom-Wrapper"
      >
        <div
          className="HomeImageSliderButtonWrapper-arrow-6"
        >
          <div
            className="Arrow-container-7"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
              data-testid="Arrow-Button-Left"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <img
                  alt="nata"
                  className="Arrow-left-8"
                  data-testid="Arrow-Button-Image"
                  src="arrow-disabled.png"
                />
              </span>
            </button>
          </div>
        </div>
        <div
          className="HomeImageSliderButtonWrapper-arrow-6"
        >
          <div
            className="Arrow-container-7"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text"
              data-testid="Arrow-Button-Right"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <img
                  alt="nata"
                  className="Arrow-right-9"
                  data-testid="Arrow-Button-Image"
                  src="arrow-enabled.png"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ImageSlider-card_slider-3"
      data-testid="ImageSlider-Card-Slider"
      onWheel={[Function]}
      style={
        Object {
          "height": "undefinedpx",
        }
      }
    >
      <div
        className="ImageSlider-card_slider_wrapper-4"
        style={
          Object {
            "transform": "translateX(-0%)",
          }
        }
      >
        <div
          className="CollectionCard-spacing_selected-10"
          data-testid="Home-Collection-Card-first-test-collection-title"
        >
          <div
            className="CollectionCard-container-12"
            data-testid="Home-Collection-Card-Container"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="gallery/first-test-collection-key"
              onBlur={[Function]}
              onFocus={[Function]}
              style={
                Object {
                  "textDecoration": "none",
                }
              }
            >
              <img
                alt=""
                className="CollectionCardImage-layout-13"
                data-testid="Collection-Card-Image"
                height={130}
                src="test-image-url"
                width="350"
              />
              <div
                className="CollectionCardText-text_wrapper-15"
                data-testid="Collection-Card-Text-Wrapper"
              >
                <img
                  alt=""
                  className="CollectionCardText-icon-16"
                  data-testid="Collection-Card-Image-Icon"
                  height={20}
                  src="collection-image-o.png"
                  width={20}
                />
                <p
                  className="CollectionCardText-text-17"
                  data-testid="Collection-Card-Title"
                >
                  first-test-collection-title
                </p>
                <div
                  className="CollectionCardText-tag-18"
                >
                  <div
                    className="HomeCollectionTag-layout-19"
                    data-testid="tag-layout"
                    style={
                      Object {
                        "background": "#323232",
                      }
                    }
                  >
                    Default
                  </div>
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          className="CollectionCard-spacing_unselected-11"
          data-testid="Home-Collection-Card-second-test-collection-title"
        >
          <div
            className="CollectionCard-container-12"
            data-testid="Home-Collection-Card-Container"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="gallery/second-test-collection-key"
              onBlur={[Function]}
              onFocus={[Function]}
              style={
                Object {
                  "textDecoration": "none",
                }
              }
            >
              <img
                alt=""
                className="CollectionCardImage-layout-13"
                data-testid="Collection-Card-Image"
                height={130}
                src="test-image-url"
                width="350"
              />
              <div
                className="CollectionCardText-text_wrapper-15"
                data-testid="Collection-Card-Text-Wrapper"
              >
                <img
                  alt=""
                  className="CollectionCardText-icon-16"
                  data-testid="Collection-Card-Image-Icon"
                  height={20}
                  src="collection-image-o.png"
                  width={20}
                />
                <p
                  className="CollectionCardText-text-17"
                  data-testid="Collection-Card-Title"
                >
                  second-test-collection-title
                </p>
                <div
                  className="CollectionCardText-tag-18"
                >
                  <div
                    className="HomeCollectionTag-layout-19"
                    data-testid="tag-layout"
                    style={
                      Object {
                        "background": "#323232",
                      }
                    }
                  >
                    Default
                  </div>
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          className="CollectionCard-spacing_unselected-11"
          data-testid="Home-Collection-Card-third-test-collection-title"
        >
          <div
            className="CollectionCard-container-12"
            data-testid="Home-Collection-Card-Container"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="gallery/third-test-collection-key"
              onBlur={[Function]}
              onFocus={[Function]}
              style={
                Object {
                  "textDecoration": "none",
                }
              }
            >
              <img
                alt=""
                className="CollectionCardImage-layout-13"
                data-testid="Collection-Card-Image"
                height={130}
                src="test-image-url"
                width="350"
              />
              <div
                className="CollectionCardText-text_wrapper-15"
                data-testid="Collection-Card-Text-Wrapper"
              >
                <img
                  alt=""
                  className="CollectionCardText-icon-16"
                  data-testid="Collection-Card-Image-Icon"
                  height={20}
                  src="collection-image-o.png"
                  width={20}
                />
                <p
                  className="CollectionCardText-text-17"
                  data-testid="Collection-Card-Title"
                >
                  third-test-collection-title
                </p>
                <div
                  className="CollectionCardText-tag-18"
                >
                  <div
                    className="HomeCollectionTag-layout-19"
                    data-testid="tag-layout"
                    style={
                      Object {
                        "background": "#323232",
                      }
                    }
                  >
                    Default
                  </div>
                </div>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
